model/convert/service_event.go

```diff
@@ -49,58 +49,159 @@ func convertServiceEventEpochSetup(event flow.Event) (*flow.ServiceEvent, error)
 		return nil, fmt.Errorf("could not unmarshal event payload: %w", err)
 	}
 
-	// parse cadence types to required fields
-	setup := new(flow.EpochSetup)
-
 	// NOTE: variable names prefixed with cdc represent cadence types
 	cdcEvent, ok := payload.(cadence.Event)
 	if !ok {
 		return nil, invalidCadenceTypeError("payload", payload, cadence.Event{})
 	}
 
-	if len(cdcEvent.Fields) < 9 {
+	expectedFieldCount := 9
+	if len(cdcEvent.Fields) < expectedFieldCount {
 		return nil, fmt.Errorf(
-			"insufficient fields in EpochSetup event (%d < 9)",
+			"insufficient fields in EpochSetup event (%d < %d)",
 			len(cdcEvent.Fields),
+			expectedFieldCount,
 		)
 	}
 
-	// extract simple fields
-	counter, ok := cdcEvent.Fields[0].(cadence.UInt64)
-	if !ok {
-		return nil, invalidCadenceTypeError(
-			"counter",
-			cdcEvent.Fields[0],
-			cadence.UInt64(0),
-		)
+	if cdcEvent.Type() == nil {
+		return nil, fmt.Errorf("EpochSetup event doesn't have type")
 	}
-	setup.Counter = uint64(counter)
-	firstView, ok := cdcEvent.Fields[2].(cadence.UInt64)
-	if !ok {
-		return nil, invalidCadenceTypeError(
-			"firstView",
-			cdcEvent.Fields[2],
-			cadence.UInt64(0),
-		)
+
+	// parse cadence types to required fields
+
+	var counter cadence.UInt64
+	var firstView cadence.UInt64
+	var finalView cadence.UInt64
+	var randomSrcHex cadence.String
+	var dkgPhase1FinalView cadence.UInt64
+	var dkgPhase2FinalView cadence.UInt64
+	var dkgPhase3FinalView cadence.UInt64
+	var cdcClusters cadence.Array
+	var cdcParticipants cadence.Array
+	var foundFieldCount int
+
+	evt := cdcEvent.Type().(*cadence.EventType)
+
+	for i, f := range evt.Fields {
+		switch f.Identifier {
+		case "counter":
+			foundFieldCount++
```

Comment:
can fields potentially have duplicate identifier (e.g., multiple "counter")?  if so, then you should track found field differently

Reply:
> can fields potentially have duplicate identifier (e.g., multiple "counter")?
No, it would be invalid Cadence event.